ARCHITECTURE.md

```diff
@@ -0,0 +1,275 @@
+# HyperDB Architecture
+
+HyperDB is a scalable peer-to-peer key-value database.
+
+## Filesystem metaphor
+
+HyperDB is structured to be used much like a traditional hierarchical
+filesystem. A value can be written and read at locations like `/foo/bar/baz`,
+and the API supports querying or tracking values at subpaths, like how watching
+for changes on `/foo/bar` will report both changes to `/foo/bar/baz` and also
+`/foo/bar/19`.
+
+## Set of append-only logs
+
+A HyperDB is fundamentally a set of
+[hypercore](https://github.com/mafintosh/hypercore)s. A *hypercore* is a secure
+append-only log that is identified by a public key, and can only be written to
+by the holder of the corresponding private key. Because it is append-only, old
+values cannot be deleted nor modified. Because it is secure, any other peer in
+the network can verify these attributes without inherently trusting the author.
+
+Each entry in a hypercore has a *sequence number*, that increments by 1 with
+each write, starting at 0 (`seq=0`).
+
+HyperDB builds its hierarchical key-value store on top of these hypercores, and
+also provides facilities for authorization, and replication of those member
+hypercores.
+
+### Directed acyclic graph
+
+The combination of all operations performed on a HyperDB by all of its members
+forms a DAG (*directed acyclic graph*). Each database operations link backward
+to all of the known "heads" in the graph.
+
+For example, let's say Alice starts a new HyperDB and writes 2 values to it:
+
+```
+// Feed
+
+0 (/foo/bar = 'baz')
+1 (/foo/2   = '{ "some": "json" }')
+
+
+// Graph
+
+Alice:  0  <---  1
+```
+
+Where sequence number 1 (the second entry) refers to sequence number 0 on the
+same feed (Alice's).
+
+Now Alice *authorizes* Bob to write to the HyperDB. Internally, this means Alice
+writes a special message to her feed saying that Bob's feed (identified by his
+public key) should be read and replicated in by other participants. Her feed
+becomes
+
+```
+// Feed
+
+0 (/foo/bar = 'baz')
+1 (/foo/2   = '{ "some": "json" }')
+2 (''       = '')
+
+
+// Graph
+
+Alice: 0  <---  1  <---  2
+```
+
+Authorization is formatted internally in a special way so that it isn't
+interpreted as a key/value pair.
+
+Now Bob writes a value to his feed, and then Alice and Bob sync. The result is:
+
+```
+// Feed
+
+//// Alice
+0 (/foo/bar = 'baz')
+1 (/foo/2   = '{ "some": "json" }')
+2 (''       = '')
+
+//// Bob
+0 (/a/b     = '12')
+
+
+// Graph
+
+Alice: 0  <---  1  <---  2
+Bob  : 0
+```
+
+Notice that none of Alice's entries refer to Bob's, and vice versa. This is
+because neither has written any entries to their feeds since the two became
+aware of each other (authorized & replicated each other's feeds).
```

Comment:
Ah - and this makes sense to me now because of your discussion of ties’ below. 

If I understand correctly - they dont need to know about each other because when a lookup is initiated the db checks the latest value in each feed. So if looking for Bobs addition it will be found immediately. No need to refer to Alice.

Is this right?

Reply:
I'm not sure that's right. Or maybe I've misunderstood? I'll try to rephrase:

Alice and Bob *know about* each other (as in, they've replicated each other's feeds), but *at the point in time that Alice and Bob wrote their latest entries to the db*, they did not know about each other. Every node contains links backwards to the heads it knew about *at write-time*.

Reply:
that is much clearer.